### PR TITLE
support gc stale peer. 

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @BusyJay @disksing @zhangjinpeng1987

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -322,24 +329,94 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+
+            if util::is_epoch_stale(from_epoch, epoch) &&
+               util::find_peer(region, from_peer.get_store_id()).is_none() {
+                warn!("raft message {:?} is stale {:?}, current {:?}, tell to gc",
+                      msg_type,
+                      from_epoch,
+                      epoch);
+                self.send_gc_peer_msg(region_id, to_peer.clone(), from_peer.clone(), epoch.clone());
+                return Ok(true);
+            }
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: return Ok(false) then Remove else branch

**Comment by ngaut** [comment]
PTAL @BusyJay

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -257,14 +257,21 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             return Ok(());
         }
 
+        if msg.has_is_tombstone() && msg.get_is_tombstone() {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: I think  just `get_is_tombstone` is enough.

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -407,6 +511,35 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         Ok(())
     }
 
+    fn destory_peer(&mut self, region_id: u64, peer: metapb::Peer) {
+        warn!("destroy peer {:?} for region {}", peer, region_id);
+        // TODO: should we check None here?
+        // Can we destroy it in another thread later?
+        let mut p = self.region_peers.remove(&region_id).unwrap();
+        // We can't destroy a peer which is applying snapshot.
+        assert!(!p.is_applying_snap());
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Maybe we can just ignore this message when applying snap.

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -166,25 +166,9 @@ impl Peer {
     // will be retrieved later after appling snapshot.
     pub fn replicate<T: Transport, C: PdClient>(store: &mut Store<T, C>,
                                                 region_id: u64,
-                                                from_epoch: &metapb::RegionEpoch,
+
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why blank line here?

Comment (siddontang) [inline src/raftstore/store/peer.rs]: my fault.

File: src/raftstore/store/store.rs (hunk 3 above)
Comment (siddontang) [inline src/raftstore/store/store.rs]: now we can't destroy region which is applying snapshot, so here just assert.

**Comment by BusyJay** [comment]
LGTM

**Comment by disksing** [comment]
LGTM

**Pull request merged**